### PR TITLE
Option for creating a task on the issue comment

### PR DESCRIPTION
**Pull request by acopet**
Linked to issue #7.

New feature to create automatically a task for all violations which reach a severity threshold.

### CODE REVIEW

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java [hunk 1]
```diff
@@ -300,4 +325,28 @@ public void resetComments(String project, String repository, String pullRequestI
       LOGGER.debug("Exception stack trace", e);
     }
   }
+  
+  /**
+   * Get available severity to create a task.
+   */
+  public List<String> getAvailableSeverities() {
```
Comment (nvillard) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: why not using a switch(treshold) {
case INFO:
  result.add(INFO);
case MINOR:
...
break;
} isntead?
More readable for me

File: src/main/java/org/sonar/plugins/stash/issue/collector/StashCollector.java [hunk 2]
```diff
@@ -57,6 +45,62 @@ public static StashCommentReport extractComments(String jsonBody) throws StashRe
     return result;
   }
 
+  public static StashComment extractComment(String jsonBody, String path, Long line) throws StashReportExtractionException {
```
Comment (nvillard) [inline src/main/java/org/sonar/plugins/stash/issue/collector/StashCollector.java]: I would recomend to factorize the common part of the 2 methods in a 3rd, not public method. Something like:
  static StashComment extractComment(JSONObject jsonComment, String path, Long line) throws StashReportExtractionException {
    StashComment result = null;
    
      long id = (long) jsonComment.get("id");
      String message = (String) jsonComment.get("text");
      
      long version = (long) jsonComment.get("version");
      
      JSONObject jsonAuthor = (JSONObject) jsonComment.get("author");
      StashUser stashUser = extractUser(jsonAuthor.toJSONString());
      
      result = new StashComment(id, message, path, line, stashUser, version);

    return result;
  }
  
  

  public static StashComment extractComment(String jsonBody, String path, Long line) throws StashReportExtractionException {
    StashComment result = null;
    
    try {
      JSONObject jsonComment = (JSONObject) new JSONParser().parse(jsonBody);
      
      result = extractComment(jsonComment, path, line);

    } catch (ParseException e) {
      throw new StashReportExtractionException(e);
    }
    
    return result;
  }
  
  public static StashComment extractComment(String jsonBody) throws StashReportExtractionException {
    StashComment result = null;
    
    try {
      JSONObject jsonComment = (JSONObject) new JSONParser().parse(jsonBody);
      
      JSONObject jsonAnchor = (JSONObject) jsonComment.get("anchor");
      if (jsonAnchor == null) {
        throw new StashReportExtractionException("JSON Comment does not contain any \"anchor\" tag to describe comment \"line\" and \"path\"");
      }
      
      String path = (String) jsonAnchor.get("path");
        
      // can be null if comment is attached to the global file
      Long line = (Long) jsonAnchor.get("line");
      
      result = extractComment(jsonComment, path, line);

    } catch (ParseException e) {
      throw new StashReportExtractionException(e);
    }
    
    return result;
  }

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java [hunk 3]
```diff
@@ -300,4 +325,28 @@ public void resetComments(String project, String repository, String pullRequestI
       LOGGER.debug("Exception stack trace", e);
     }
   }
+  
+  /**
+   * Get available severity to create a task.
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: typo: severities

File: src/main/java/org/sonar/plugins/stash/StashPlugin.java [hunk 4]
```diff
@@ -21,7 +23,12 @@
...
 
-  private static final String CONFIG_PAGE_SUB_CATEGORY_GENERAL = "General";
+  private static final String CONFIG_PAGE_SUB_CATEGORY_STASH = "Stash";
+  
+  public static final String SEVERITY_NONE = "NONE";
+  
+  // INFO, MINOR, MAJOR, CRITICAL, BLOCKER
+  protected static final List<String> SEVERITY_LIST = Severity.ALL;
   
   public static final String INCREMENTAL_MODE = "incremental";
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/StashPlugin.java]: Can be removed

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java [hunk 5]
```diff
@@ -300,4 +325,28 @@ public void resetComments(String project, String repository, String pullRequestI
...
+      String threshold = config.getTaskIssueSeverityThreshold();
+        
+      // threshold == NONE, no severities available 
+      if (! StringUtils.equals(threshold, StashPlugin.SEVERITY_NONE)) {
+        
+        // INFO, MINOR, MAJOR, CRITICAL, BLOCKER
+        boolean hit = false;
+        for (String severity : StashPlugin.SEVERITY_LIST) {
+          
+          if (hit || StringUtils.equals(severity, threshold)) {
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: Possible discrepancy: while the severity is not matched, then we add the severity to result and we carry on.
For example if it's MAJOR, then we should have result to [INFO,MINOR,MAJOR]
Proposal:
// INFO, MINOR, MAJOR, CRITICAL, BLOCKER
        for (String severity : StashPlugin.SEVERITY_LIST) {
          result.add(severity);
          if (StringUtils.equals(severity, threshold)) {
            break;
          }
        }

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java [hunk 6]
```diff
@@ -300,4 +325,28 @@ public void resetComments(String project, String repository, String pullRequestI
...
+   */
+  public List<String> getAvailableSeverities() {
+      List<String> result = new ArrayList<>();
+      String threshold = config.getTaskIssueSeverityThreshold();
+        
+      // threshold == NONE, no severities available 
+      if (! StringUtils.equals(threshold, StashPlugin.SEVERITY_NONE)) {
+        
+        // INFO, MINOR, MAJOR, CRITICAL, BLOCKER
+        boolean hit = false;
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: for readability perh

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java (hunk 1 above)
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: Suggestion renaming: getAvailableSeverities => getReportedSeverities

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java [hunk 7]
```diff
@@ -289,7 +302,19 @@ public void resetComments(String project, String repository, String pullRequestI
         
         // delete comment if published by the current SQ user
         if (sonarUser.getId() == comment.getAuthor().getId()) {
-          stashClient.deletePullRequestComment(project, repository, pullRequestId, comment);
+          
+          // comment contains tasks which cannot be deleted => do nothing
+          if (comment.containsNotDeletableTasks()) {
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: Suggestion: invert the logic
if (comment.containsDeletableTasks() ) {
  for (StashTask ...
... else
  LOGGER.debug(....

File: src/main/java/org/sonar/plugins/stash/issue/StashComment.java [hunk 8]
```diff
@@ -53,6 +63,27 @@ public long getVersion() {
...
   
+  public List<StashTask> getTasks() {
+    return tasks;
+  }
+  
+  public void addTask(StashTask task){
+    tasks.add(task);
+  }
+  
+  public boolean containsNotDeletableTasks() {
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/issue/StashComment.java]: Suggestion: avoid negation
public boolean containsDeletableTasks() {
    boolean result = true;

    for (StashTask task: tasks){
      if (! task.isDeletable()){
        result = false;
        break;
      }
    }

    return result;
  }

File: src/main/java/org/sonar/plugins/stash/issue/StashTask.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package org.sonar.plugins.stash.issue;
+
+public class StashTask {
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/issue/StashTask.java]: Readability: Might be good to add the vocation of the class in comment.

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java (hunk 1 above)
Comment (acopet) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: getReportedSeverities new name of the method.

File: src/main/java/org/sonar/plugins/stash/StashPlugin.java (hunk 4 above)
Comment (acopet) [inline src/main/java/org/sonar/plugins/stash/StashPlugin.java]: Done!

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java (hunk 7 above)
Comment (acopet) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: Ambiguity probably comes from wrong naming of the method "containsNotDeletableTasks".
This method will be renamed as "containsPermanentTasks" to make more readable this code.

File: src/main/java/org/sonar/plugins/stash/issue/StashComment.java (hunk 8 above)
Comment (acopet) [inline src/main/java/org/sonar/plugins/stash/issue/StashComment.java]: Ambiguity probably comes from wrong naming of the method "containsNotDeletableTasks".
This method will be renamed as "containsPermanentTasks" to make more readable this code.

File: src/main/java/org/sonar/plugins/stash/issue/collector/StashCollector.java (hunk 2 above)
Comment (acopet) [inline src/main/java/org/sonar/plugins/stash/issue/collector/StashCollector.java]: Code has been refactored in two methods.

File: src/main/java/org/sonar/plugins/stash/issue/StashTask.java (hunk 9 above)
Comment (acopet) [inline src/main/java/org/sonar/plugins/stash/issue/StashTask.java]: I would prefer to keep this simple class like this: it finally just a class to describe a Stash task.

**Comment by acopet** [comment]
@mathieusimon, @nvillard 
**All pull-request comments has been integrated.**

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java [hunk 10]
```diff
@@ -300,4 +325,28 @@ public void resetComments(String project, String repository, String pullRequestI
       LOGGER.debug("Exception stack trace", e);
     }
   }
+  
+  /**
+   * Get available severities to create a task.
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: Get reported severities...

File: src/main/java/org/sonar/plugins/stash/StashRequestFacade.java [hunk 11]
```diff
@@ -300,4 +325,28 @@ public void resetComments(String project, String repository, String pullRequestI
...
+   */
+  public List<String> getReportedSeverities() {
+      List<String> result = new ArrayList<>();
+      String threshold = config.getTaskIssueSeverityThreshold();
+        
+      // threshold == NONE, no severities available
```
Comment (mathieusimon) [inline src/main/java/org/sonar/plugins/stash/StashRequestFacade.java]: ..., no severities to report

File: src/test/java/org/sonar/plugins/stash/StashRequestFacadeTest.java [hunk 12]
```diff
@@ -400,4 +531,39 @@ public void testAddPullRequestReviewerWithReviewerAlreadyAdded() throws Exceptio
     
     verify(stashClient, times(0)).addPullRequestReviewer(STASH_PROJECT, STASH_REPOSITORY, STASH_PULLREQUEST_ID, (long) 1, reviewers);
   }
+  
+  @Test
+  public void testGetAvailableSeverities() {
```
Comment (mathieusimon) [inline src/test/java/org/sonar/plugins/stash/StashRequestFacadeTest.java]: In order to be iso everywhere change for testGetReportedSeverities

**Comment by mathieusimon** [comment]
Ok for me!

**Comment by nvillard** [comment]
Fine with me

**Pull request merged**